ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -588,6 +597,68 @@ static PrjFS_Result HandleEnumerateDirectoryRequest(const MessageHeader* request
     return callbackResult;
 }
 
+static PrjFS_Result HandleRecursivelyEnumerateDirectoryRequest(const MessageHeader* request, const char* path)
+{
+#ifdef DEBUG
+    std::cout << "PrjFSLib.HandleRecursivelyEnumerateDirectoryRequest: " << path << std::endl;
+#endif
+    
+    DIR* directory = nullptr;
+    PrjFS_Result result = PrjFS_Result_Success;
+    std::list<std::string> directoryRelativePaths;
+    directoryRelativePaths.push_back(path);
+    
+    // Walk each directory, expanding those that are found to be empty
+    char pathBuffer[PrjFSMaxPath];
+    std::list<std::string>::iterator relativePathIter = directoryRelativePaths.begin();
+    while(relativePathIter != directoryRelativePaths.end())
+    {
+        CombinePaths(s_virtualizationRootFullPath.c_str(), relativePathIter->c_str(), pathBuffer);
+    
+        // TODO(Mac): how should we handle scenarios where we were unable to fully expand the directory and
```

Comment:
> what will happen in the scenario where a directory's recursive children are partially on disk already

In your example, assuming we get `KAUTH_VNODE_DELETE` for `A` first, the directories would be walked breadth-first:

1. Push `A` into `directoryRelativePaths`
2. Grab the next entry in `directoryRelativePaths` (i.e. `A`), it does not need to be expanded
3. Walk all children of `A` and push any directories (i.e. `B` and `C`) onto `directoryRelativePaths`
4. Grab the next entry in `directoryRelativePaths` (i.e. `B`), it does not need to be expanded
5. Walk all children of `B`, there are no directories
6. Grab the next entry in `directoryRelativePaths` (i.e. `C`), it needs to be expanded, and so expand it
7. Walk all children of `C`, there are no directories
8. There are no more entries in `directoryRelativePaths`

(Note that writing this out I realize we should probably just pull the entry off the front of `directoryRelativePaths` after we process it rather than leaving it in `directoryRelativePaths`)

> How does that behave normally?

It's possible that without my changes we'd already be expanding everything (because the delete is recursive).  I can test that out and confirm.  

The real reason we need this change is for rename operations, as they will not walk the tree recursively.

> How does it behave if some of the expansions fail, especially with more complicated data shapes than my example?

I will test out delete specifically and see in what order `KAUTH_VNODE_DELETE` is received.

At the end of the day though, if any of the expansions fail the kext will return `DENY` and prevent the delete\rename from proceeding.  I can force some failures to occur and reply back with what I see.

Assuming that each subfolder isn't marked as expanded until all of its placeholders are written, then (as mentioned in the earlier comment), these changes aren't introducing much new complexity.  If the delete is requested again the user-mode library will skip over the folders that were expanded in the previous attempt.